wsrep_tls_service.h

```diff
@@ -0,0 +1,357 @@
+ */
+typedef struct wsrep_tls_stream_st
+{
+    /**
+     * File descriptor corresponding to the stream. The provider is
+     * responsible in opening and closing the socket.
+     */
+    int fd;
+    /**
+     * Pointer to TLS context. For initializing valid TLS context see
+     * wsrep_tls_context_create_t below.
+     */
+    wsrep_tls_context_t* context;
+    /**
+     * Opaque pointer reserved for application use.
+     */
+    void* opaque;
+} wsrep_tls_stream_t;
+
+
+
+/**
+ * Create a new TLS context.
+ *
+ * This function will be called by the provider before the
+ * stream functionality is used. The application must return
+ * non-NULL pointer to the context and keep it alive until
+ * wsrep_tls_context_free_t is called.
+ *
+ * @return Pointer to TLS context.
+ */
+typedef wsrep_tls_context_t* (*wsrep_tls_context_create_t)();
+
+/**
+ * Free TLS context.
+ *
+ * This function is called by the provider when the stream functionality
+ * is shut down during provider deinitialization. When this function
+ * is called, the application is free to destroy resources allocated
+ * for the TLS context.
+ *
+ * @param tls_context TLS context to be freed.
+ */
+typedef void (*wsrep_tls_context_free_t)(wsrep_tls_context_t* tls_context);
+
+/*
+ * Stream operations.
+ */
+
+/**
+ * Enumeration for return codes.
+ */
+enum wsrep_tls_result
+{
+    /**
+     * The operation completed successfully, no further actions
+     * are necessary.
+     */
+    wsrep_tls_result_success = 0,
+    /**
+     * The operation completed successfully, but the application side wants
+     * to make further reads. The provider must wait until the stream
+     * becomes readable and then try the same operation again.
+     */
+    wsrep_tls_result_want_read,
+    /**
+     * The operation completed successfully, but the application side wants
+     * to make further writes. The provider must wait until the stream
+     * becomes writable and then try the same operation again.
+     */
+    wsrep_tls_result_want_write,
+    /**
+     * End of file was read from the stream. This result is needed to
+     * make difference between graceful stream shutdown and zero length
+     * reads which result from errors.
+     */
+    wsrep_tls_result_eof,
+    /**
+     * An error occurred. The specific error reason must be
+     * queried with wsrep_tls_stream_get_error_number and
+     * wsrep_tls_stream_get_error_category.
+     */
+    wsrep_tls_result_error
+};
+
+/**
+ * Initialize a new TLS stream.
+ *
+ * Initialize the stream for IO operations. During this call the
+ * application must set up all of the data structures needed for
+ * IO, but must not do any reads or writes into the stream yet.
+ *
+ * @param context TLS context.
+ * @param stream TLS stream to be initialized.
+ *
+ * @return Zero on success, system error number on error.
+ */
+typedef int (*wsrep_tls_stream_init_t)(
+    wsrep_tls_context_t* context,
+    wsrep_tls_stream_t* stream);
```

Comment:
It looks like all operations below happen with  `wsrep_tls_stream_t`. Why then would we need
`wsrep_tls_context_create_t` and `wsrep_tls_context_free_t` above? Wouldn't it be enough to just do
```
wsrep_tls_stream_t stream = { fd, NULL, NULL };
int ret = wsrep_tls_stream_init(&stream);
```
?

Reply:
I think we need to have some kind of pointer to application side context to avoid requirement for having application side global state/variable for the TLS context. However, now I think the way of creating it with calls to context create/free is fundamentally wrong, this context should be provided when the provider side service is initialized via `wsrep_init_tls_service_v1()` call.